WIP: Pre-review for v1.3.0 release

ankostis opened the pull request:
Continuation of #30, containing review-fixes in huyng/graphkit#1.

- ENH: heavily reinforced exception annotations ("jetsam"):
  - FIX: (8f3ec3a) outer graphs/ops do not override the inner cause.
  + ENH: retrofitted exception-annotations as a single dictionary, to print it in one shot 
    (8f3ec3a & 8d0de1f) 
  - enh: more data in a dictionary 
  + TCs: Add thorough TCs (8f3ec3a & b8063e5).
- REFACT(net): rename `Delete`-->`Evict`, removed `Placeholder` from nadanodes, privatize node-classes.
- ENH(sideffects): make them always DIFFERENT from regular DATA, to allow to co-exist.
- TCs: `pytest` now checks sphinx-site builds without any warnings. 
- FIX(plan): multithreading was broken due shared `plan.executed`: moved to new class `Solution`. 
- [ ] `graphop` in docs
- [ ] typo(test): overriden-->overriDDen
- [ ] DROP sideffects

ankostis committed:
FEAT: +TOKEN modifier for side-effect functions

Tokens work as usual while solving the DAG but
they are never assigned any values to/from the operation's functions.

+ TC included.
+ Docs updated.\+ Added `modifiers` superclass to facilitate identification code.
+ refactored FunctionalOperation._compute().

ankostis committed:
fix(func): 1-item outs broke with token modifiers...

+ _norm_kwargs:
  + use isinstance() instead of type-equality checks,
    to support new modier classes;
    + avoid excessive dict searches with local vars.

ankostis committed:
FIX(build): py2 needs pinning networkx-2.2

ankostis committed:
refact(token): undo adding of new `modifier` class...

not really needed, better be explicit which modifier is searched.

ankostis committed:
enh(token.TC): test 2 reversed token workflows

ankostis committed:
ENH(net,#18): ignore UN-SATISFIABLE operations with partial inputs

+ The x2 TCs added just before are now passing.

ankostis committed:
doc(#13): sample code to plot workflow diagram in intro

ankostis committed:
FIX(#13): BUG in plot-diagram writtin from PY2 era,

were writing in text-mode in PY3. and failing as encoding error.

ankostis committed:
enh(plot): provide help msg on supported file-exts

ankostis committed:
TEST(plot,ci): test plotting; pip install extras in Travis

ankostis committed:
fix(plot): don't create file on unsupported formats

thanks to @andres-fr.

ankostis committed:
enh(plot.TC): expose supported writers and TC on them

ankostis committed:
ENH(OPTIONAL): optionals per OPERATION, not to NEEDs...
o
+ Closes #17

ankostis committed:
enh(test): + x2 TC breaking UNSATISFIED operations...

receiving partial inputs, needed for other operations.

ankostis committed:
refact(net): drop old `dag` nx-package

ankostis committed:
ENH(core): ORDERED SETs for DETERMINISTIC results

NOTE dict are not deterministic in <PY3.6.
So this commit would not improve determinism
in those pythons.

+ build: add `boltons` dependency for ndexedSet.
+ doc: mark all set usage if affect determinism.
+ e.g. see reproducibility problem in #14.

ankostis committed:
merge UNASTIFIABLE + ORDERED_SETs

ankostis committed:
refact(net): simpilify del-instruction loop

ankostis committed:
REFACT(unsatisfied): doubly-recursive func --> loop on topo-sorted

ankostis committed:
test(dag,#25): FAILing TC for overriding intermediate data

ankostis committed:
test(dag,#24): FAILing TC for over-pruning inetermediates when outs asked

ankostis committed:
MERGE (prune_unsatified, ordered_sets) into fix-pruning ...

needed to refactor the new pruning algorithm.

- expected 2 TCs fail for yet-to-be-solved #24 & #25 bugs.

ankostis committed:
DOC(net): explain DAG solution & compilation...

the later described in #21.

ankostis committed:
TEST(prune): +Failing x2 TCs multi-out must run but not...

override intermediate data.

More changes only for newer pruning TCs:
+ refact(test): rename graph-->netop vars for results of compose(),
  to avoid of `graph.net.graph`.
+ Explain failure modes in v1.2.4 & this merged branch (#19 + #23).

ankostis committed:
refact(dag): call compile() before compute.compute...

not after compose().

+ All TCs pass ok.
+ NOTE this is not yet what is described in #21.

ankostis committed:
test(dag): +TC checking DeleteInst vary when inputs change

ankostis committed:
doc(net): explain new DAG SOLUTION

ankostis committed:
REFACT(NET) COMPILE+COMPUTE...

+ Read the next doc-only commit to understand changes.
+ Renamed:
  + net.steps --> net.execution_plan.
  + (old)compile() --> _build_execution_plan()
  + _find_necessary_steps() --> (new)compile() + _solve_dag()
    compile() became the master function invoking _solve_dag &
    _build-execution_plan(), and do the caching.
+ refact(compute()): extract common tasks from sequential/parallel.
+ refact show_layers() to allow full-print, geting also string
  (not just printing), and using custom classes for representation.
+ Raise AssertionError when invalid class in plan.
  it's a logic error, not a language type-error.

ankostis committed:
ENH(net): move compile() after SOLVE DAG ...

to pass +TC checking DeleteInst vary when inputs change.

- x4 TCs still failing, and need revamp of dag-solution.

ankostis committed:
FIX(net): new Ops invalidate execution-plan cache...

Probaly unreported bug in v1.2.4 for '_neccessary_steps_cache`.

ankostis committed:
ENH(DAG): NEW SOLVER

+ Pruning behaves correctly also when outputs given;
  this happens by breaking incoming provide-links
  to any given intermedediate inputs.
+ Unsatisfied detection now includes those without outputs
  due to broken links (above).
+ Remove some uneeded "glue" from unsatisfied-detection code,
  leftover from previous compile() refactoring.
+ Renamed satisfiable --> satisfied.
+ Improved unknown output requested raise-message.
+ x3 TCs PASS, x1 in #24 and the first x2 in #25.
- 1x TCs in #25 still FAIL, and need "Pinning" of given-inputs
  (the operation MUST and MUST NOT run in these cases).

ankostis committed:
WIP/FIX(prune,#26): PIN intermediate inputs if operation before must run

- WIP: PARALLEL execution not adding PINS!
+ Insert "PinInstructions" in the execution-plan to avoid overwritting.
+ Add `_overwrite_collector` in `compose()` to collect re-calculated values.
+ FIX the last TC in #25.

ankostis committed:
enh(build): replace numpy with pytest...

numpy was used just for its assert_raise

ankostis committed:
feat(build): add pip-extras [test]

ankostis committed:
enh(build,ci): use pytest in travis

ankostis committed:
REFACT(net): part 3 of new dag-solver & pin refactoring

- STILL buggy PIN on PARALLEL, 2 DISABLED TCs FAIL:
  - test_pruning_with_given_intermediate_and_asked_out()
  - test_unsatisfied_operations_same_out()
+ move check if value in asked outputs before cache-evicting it
  in build-execution-plan time - compute methods
  don't need outputs anymore.
+ test: speed up parallel/multihtread TCs
  by reducing delays & repetitions.
+ refact: network rightfully adopted stray functions
  for parallel processing - they all worke on the net.graph,
+ upd: networkx api by indexing on `dag.nodes` views.
+ enh: add log message when deleting in parallel
  (in par with sequential code).
+ refact: var-renames, if-then-else simplifications, pythonisms.
+ doc: A lot!

ankostis committed:
WIP/FIX(PIN): PARALLEL DELs decide on PRUNED-dag (not full)...

- WIP: x4 TCs FAIL and still not discovered th bug :-(
+ BUT ALL+AUGMENTED PARALLEL TCs pass
  (#26 were failing some)
+ refact: net stores also `pruned_dag` (not only `steps`).
+ refact: _solve_dag() --> _prune_dag().
+ doc: +a lot.
+ TODO: store pruned_dag in own ExePlan class.

ankostis committed:
WIP/+ExecPlan class...

- SAME x4 TCs FAIL like parent.
+ refact: revive net.steps --> net.last_plan
+ BREAK: inverse the order of outputs/inputs args in net.compute()
+ refact: use iset() to merge.

ankostis committed:
enh(CI): +PY3.6 where dicts are stable

ankostis committed:
fix(plot): NetOp did not return pydot instance

ankostis committed:
refact(plot): extract plot function out of Network class...

to use it also on stary DAGs.  Keep delegation.

ankostis committed:
fix(plot): matplotlib plot was failing  in PY3 due IO io misuse

ankostis committed:
FEAT(plot): overlay Execution STEPS on diagrams

ankostis committed:
ENH(plot): +inputs, +outputs, +solution modify plotting

(see #13 for an example):

ankostis committed:
refact(plot.TC): move plot tests to early beggining

ankostis committed:
doc(plot): tell supported formats in doctest, +TC

ankostis committed:
doc(plot): add legend & example; docstring in netop.plot()

ankostis committed:
enh(plot): mark optional "needs"

ankostis committed:
ENH(plot): visual enhamcents on nodes & edges

ankostis committed:
test(plot): enhance plot test to try all #13 features;

+ test all chained plot() methods from netop.

ankostis committed:
test(optional): +x1 TC to check selective optionals with same out...

+ enh TC with parallel.
+ Scavenged from #20.

ankostis committed:
doc(plot); explain also params in user-facing API

ankostis committed:
FIX(PLOT.TC): TC was always testing PNG, ...

+ retorfitted to try all available formats.
+ list of forbidden formats based on my failres

ankostis committed:
fix(plot): don't require Matplotlib if no Window asked

ankostis committed:
test(plot): check also matplotlib show=True

ankostis committed:
ENH(plot): return SVG rendered in JUPYTER, ...

+ doc: rename in sample code: netop --> pipeline.
+ enh(build): add `ipython` in test dependencies.
+ include it in the plot TC.

ankostis committed:
refact(plot.TC): avoid writting multiple temp-files

ankostis committed:
MERGE PYTEST into PLOT to parametrize TCs.

ankostis committed:
fix(build): reuse dependencies definitions

ankostis committed:
REFACT(plot.TC): PYTESTize and parametrize

ankostis committed:
REFACT(PLOT): MOVE PLOT in own module

ankostis committed:
DROP PY3.4 - add PY3.6, PY3.7...

...pytest has problems with 3.4.

ankostis committed:
refact(plot): separate graphviz building from IO

ankostis committed:
MERGE ENHANCED-PLOTS into REFACT-ExecutionPlan to debug issues

ankostis committed:
DROP(net): list/show layers not needed, repr() is ok

ankostis committed:
ENH(plot,net): +plot() on ExecPlan; +PlotMixin ...

to avoid copying plot() doc+sig around

ankostis committed:
enh(plan): repr()

ankostis committed:
refact(plot): reorder formats-list with io

ankostis committed:
FIX(plot): distinguish Del/Pin cmds; +choice utils

- still x4 TCs fail...

ankostis committed:
refact(plot): inline imports, not to cycle with base...

to bring back PlotMixin to change together stuff.

ankostis committed:
refact(plot): move PlotMixin base-->plot module to group edit

ankostis committed:
enh(plot): +`title` arg at the bottom

ankostis committed:
FIX(plot): failing if steps not a list/ is none

ankostis committed:
ENH(plan,plot): executed operations drawn as filled

+ enh(plan): also sequential execution collects accounts executed.
+ refact(plan): executed_nodes-->executed.

ankostis committed:
FIX(net): revived last_plan was never set (HEAD~13: 64e0028)

ankostis committed:
refact(plot): reodred build-dot utils above use

ankostis committed:
doc(plot): move centrally all API doc on PlotMixin; Style

ankostis committed:
FEAT(plot,plan): +yellow broken links on original graph

ankostis committed:
refact(plot): PlotMixin --> Plotter, _plotter() --> _plot()

ankostis committed:
TEST(PLOT): Check also ExecPlan

ankostis committed:
REFACT(PLOT): Plotter builds dot & renders...

... based on Law of Demeter simplify kwargs,
defined in one place.

+ enh: plotters "suggest" kwargs, possinly to override them
  (no duplcate arg when attempted).

ankostis committed:
ENH(plan.polt): CLUSTER pruned nodes

ankostis committed:
FIX(plot): VISUAL fixes & COLOR-palettes:

+ FIX: apply thickness as in legend also for operators.
+ enh: don't cluster if no nodes pruned.
+ enh: netop includes its name a graph-title.
+ color palette: wheat filled nodes.
+ reuse common func.

ankostis committed:
ENH(netop): mark all its needs as OPTIONAL

ankostis committed:
FIX(NET): were FORGETTING PRUNED ASKED-OUTPUTs...

... bugged in the opening commit d403783 of this PR, and
discovered 68(!) commits later, and all that time had to live
with x4 broken TCs with asked-outputs.

ankostis committed:
ENH(plot,TC): pipeline delegates to last_plan, if exists

ankostis committed:
FIX(plot.TC): formats-TC were testing dot-file, not rendered...

+ ENH: strongly refatctored TC to detect nulls & dupes.

ankostis committed:
enh(net): prune also isolate data

ankostis committed:
FIX(<PY3.5): ORDERED DiGRAPH for old Python to fix TCs

+ Partial fix deterministic results (#22-2.4.3i).

ankostis committed:
ENH(plot): switch oval<-->circle ops/netops

ankostis committed:
feat(plot): +legend & TC

ankostis committed:
DOC(PLOT): +sample GRAPH & LEGEND into README...

+ test: some refacts on vars & for the uploaded images.
+ MARK SPURIOUS FAILS in < PY3.6 due to unordered dicts
  eg https://travis-ci.org/ankostis/graphkit/jobs/594813119

ankostis committed:
DROP(plot) jupyter kw, simpler to monkeypath Dot class...

to support _repr_svg(), and it works without any flag.

ankostis committed:
Rebrand TOKENS --> SIDEFFECTS

ankostis committed:
FIX(sideffect.TX): old TC was not with unsatisfied in mind,

needed to give the sidefeect as input.

ankostis committed:
FIX(plot): LEGEND mistakes, SVGize, egg graphs, ...

+ graphop label renames,
+ DOT shortening.

ankostis committed:
FIX(DOC): don't use `graph` in sample code, crash DOT, +more ...

+ FIX: `graph` is a DOT keyword, eg  pydot/pydot#111
  + Replaced `graph` --> `graphop`, inline with future refactoring.
+ Refactored example code.
+ New "Plotting" section under Quickstart.
+ Copy README instructions to main docs.
+ Abandoned `example_graph.svg`, replaced with `intro.svg`.
+ Recommend `.png` --> `.svg`, to save storage space.
- discovered BUG in MERGE saumple code (empty!).

ankostis committed:
chore(git): .gitignore plot images in root folder

ankostis committed:
chore(TCs):  mark SLOW tests, when in hurry, +`setup.cfg`

... to adopt pytest args
+ mark wheel a universal.

ankostis committed:
FEAT(TCs): +check DOCTESTs ...

+ feat: pytest-doctest configs in `setup.cfg`.
+ enh(travis): run doctests only in latest python 3.7.
+ Enabled & FIX doctests in:
  + in README
  + in docs/sources
  + in code

ankostis committed:
FIX(TCs): MERGE TCs were not ASSERTING...

got values from v1.2.4
- ALL MERGE TCs FAIL!

ankostis committed:
FIX(MERGE): broken by NEW_DAG_SOLVER (#26 ...

many commits ago.

Never got it bc TC were not checking merges!

ankostis committed:
MERGE NEW_SOLVER + PLOT --> SIDEFFECTS + OPTS in DAG

ankostis committed:
fix(TCs): enable assertions in compatibility TCs

ankostis committed:
FIX(func, TC): operation __repr__ fails with partial args...

hindering debugging sessions.

+ fix: use gettatr in both Operation & operation().
+ enh: +helpfull err-msg when forgetting to set func on Operation.
+ test: +TC to check Opetation _-repr__ in new test-file.

ankostis committed:
enh: make repr(ExecPlan) sorter

ankostis committed:
ENH(net): annotate OP-EXCEPTIONS with ExecPlan...

to aid debug sessions, inspired by @syamajala's 309338340.

+ enh: centralize calling of callables from both parrallel/serial.
+ DOC: +new sesction in "compose" with doctests exemplifying errors.

ankostis committed:
enh(travis): pytest -v implied(~) when many test-files

ankostis committed:
DOC: sphinx fixes and +INTERNALS section

+ Sphinx builds CLEAN.

ankostis committed:
fix(build): +pytest-sphinx plugin

ankostis committed:
enh(build): set README as PyPi landing page

ankostis committed:
drop(site): rtd_theme is the default for sphinx

ankostis committed:
doc(changes): +issues in v1.2.4 (old) release by yahoo

ankostis committed:
feat(site): include CHANGES sections

ankostis committed:
doc(changes): +FLOWCHART dot-file for v1.2.4

ankostis committed:
doc(changes): +FLOWCHART-v1.3.0

ankostis committed:
enh(site): sphins-extlinks extension for GH-issue links

ankostis committed:
DOC(changes): v1.2.4

ankostis committed:
chore: add myself to the authors

ankostis committed:
DOC: update package coordinates in `setup.py`

ankostis committed:
DOC(BADGES): more of them

ankostis committed:
doc(changes): mark dates of all old pypi-releases.

ankostis committed:
chore(ver): bump version 1.2.4-->1.3.0

ankostis committed:
minor leftovers (refact & docs)

ankostis committed:
STYLE: BLACKen code format

ankostis committed:
DOC(api): automodules

ankostis committed:
enh(api): import also modifiers from base package

ankostis committed:
doc: add project coords in package __strings__

ankostis committed:
fix(TC): pytest mark xfail(PY<36) due to unstable dicts

ankostis committed:
enh(ops): __repr__ listify needs/provides...

so as to politely print IndexedSets in those attributes.

ankostis committed:
enh(modifiers): no __slots__, repr() tell their class

ankostis committed:
REVERT(#20, 94b7291): not asking edges for OPTIONALs ...

but still annotate edges with optional edge data-attribute.

Reverted bc Operation must not know its network, to belong to more than one.
Also the `Operation.net` contradicted `NetwotkOperation.net`;
the later indeed is the network it wraps
(not the net it is part of).

ankostis committed:
fix(op): syntax when bad parallel-method choice given

ankostis committed:
doc(CHANGES): backport history from GitHub-releaes

ankostis committed:
fix(doc): reorder chapters, changes at the bottom, ...

+ homegine titles
+ fix: autodoc renders nothing without :special-members:

ankostis committed:
ENH(plot): show SIDEFFECTS on diagrams, MORE
+ enh(net): mark dag-edges with sideffects;
+ enh(plot): plot sideffect inks as such;
+ enh(plot): update legend;
+ doc(plot): move legend text along with the code producing it.
+ doc: stray image fixes.

ankostis committed:
FIX(PyPi): +TC to check if REAME valid markdown, ...

+ fix glitches in README.

ankostis committed:
feat(build): add build.sh script because ...

if built without cleaing, stray artifacts might make it into wheel.

ankostis committed:
DOC: improved  Opening of the Project

ankostis committed:
fix(pytest): list test items or --ignore=setup.py not working....

and when giving `pytest --lf` i get:

_____________________________________ ERROR collecting setup.py _____________________________________
/usr/lib/python3.7/distutils/fancy_getopt.py:233: in getopt
    opts, args = getopt.getopt(args, short_opts, self.long_opts)
/usr/lib/python3.7/getopt.py:93: in getopt
    opts, args = do_longs(opts, args[0][2:], longopts, args[1:])
/usr/lib/python3.7/getopt.py:157: in do_longs
    has_arg, opt = long_has_args(opt, longopts)
/usr/lib/python3.7/getopt.py:174: in long_has_args
    raise GetoptError(_('option --%s not recognized') % opt, opt)
E   getopt.GetoptError: option --lf not recognized

ankostis committed:
fic(doc): CODE in README was broken unbackquoted

ankostis committed:
ENH(ops): ANNNOTATE func-ERRORS with internal data,..

as debug aid (was doing that for plans already).

+ enh: annotate also networks.

ankostis committed:
FIX(doc): README were not PyPi-validating

ankostis committed:
doc(plot): explain how to reset plot to bare-bone

ankostis committed:
DOC(plot): MOVE plot + debug sections to own CHAPTER

ankostis committed:
refact(doc): simplify composition.rst name

ankostis committed:
DOC(PLOT): put BAREBONE in the top

ankostis committed:
refact(net): REORDER SOURCES

ankostis committed:
REFACT(net): Delete-->Evict, -Placeholder, privatize node-classes

ankostis committed:
FIX(plot): unquoted `graph` labelled graphs crash pydot/pydot#111

ankostis committed:
fix(doc): 2nd full-plot in README not matching sample code

ankostis committed:
fix(doc.TC): +2 index.rst doctests were just comments

ankostis committed:
test(net): check Pin & Evict combine appropriately

ankostis committed:
ENH(JETSAM): generify EX-ANNOTATION mechaism ...

+ FEAT(base): `jetsam()` machinery to colect from locals().
+ enh(net): jestam in compute().
+ enh: shorten annotation-names.
+ doc(net): explain why eviction cannot clan all solution
  (drop old todo).
+ refact(func): homogenize `resultS` in _compute() with other ops.
+ DROP(DOC): cloned debugging sections were
  not removed from composition.rst.
+ doc(debug): minot enhancements.

ankostis committed:
FEAT(TCs): check SITE and fail on errors

ankostis committed:
FEAT(TCs): check SITE and fail on errors, but...

FAILs due to missing image in composition.rst.

ankostis committed:
fix(site): travs pass now with new site check

ankostis committed:
enh(travis): coverage also from older pythons

ankostis commented:
@huyng can you merge #30 into master (partially reviewed in huyng/graphkit#1), to trim this long commit-history?

ankostis committed:
refact(TC): import directly modifiers (shorter code)

ankostis committed:
ENH(netop): accept SINGLE STR as asked-outs ...

... shortcut instead of demanding the usual singular lists.

+ change some sample calls in the docs.

ankostis committed:
enh(jetsham): allow *vars as 1-1 mappings; rename jetsams; ...

+ ENH(net):add forgotten `outputs` jetsam earlier in compute().
+ refact: Had to rename JETSAMs to take adavntage of 1-1 mappngs.
+ refact: updated jetsam usage to reduce source-lines.
+ renamed jetsam(arg-names).

ankostis committed:
test(net): check compute()-msg when bad outs asked

ankostis committed:
FIX(net.sideffects): typo in add_op() were mixing needs/provides  ...

+ refact(net): use same loop-var in add_op(),
  to avoid copy-paste mistakes, like above.

ankostis committed:
enh(net): +repr() listing graph-nodes

ankostis committed:
FIX(DAG): broken_dag had PLAIN-STR instead of DataNode...

bc subgraph was taken on plain string outputs.

+ minor upd err-msg terminology.

ankostis committed:
fix(jetsam): PY2 **kw syntax

ankostis committed:
ENH(sideffects): DIFFER from regular DATA...

to allow the same name to be used as regular and sideffect data node.

+ ENH: sideffect-strclass now contains its class-name.
+ DOC: a lot, comply with docstrings 1-liners;
  changes also on optionals.
+ ENH(TCs): +x2 elaborate TCs, check SEs differ from data.

ankostis committed:
test(jetsam): check actual methods using it

ankostis committed:
refact(jetsam): simpler API no contextlib...

but cannot etablish call correctnes, relying on recetn TCs.

huyng commented:
I haven't merged the previous commits because they introduced that multithreading [bug](https://github.com/huyng/graphkit/pull/1#pullrequestreview-300440464).

This PR is growing quite large and is combining several issues. 

It would be better to break this up into several PRs:

1.  fixes for #25 and #24 while not introducing the multithreading bug
2. additional plotting & debugging niceties as long as it doesn't introduce too much complexity into the code or reduce performance of graph execution.

ankostis commented:
The fix for #24 & #25 is #26.
Please merge this asap.

ankostis commented:
I will try to fix the multithreading bug here, and backport to #26.

huyng commented:
PR #26 is still causing the multithreading error from the following code. If I merge the code, it will break our master branch, which I'm trying to avoid here. We'll have to wait until the multithreading issue is resolved. I'd  recommend that you work off of your  PR#26's branch to resolve the multithreading issue before moving forward with other changes as it will make the review process easier to follow (since the PR deals primarily with the issues #24 and #25). 

```
from multiprocessing.dummy import Pool
from operator import mul, sub
from graphkit import compose, operation

# Computes |a|^p.
def abspow(a, p):
    c = abs(a) ** p
    return c

# Compose the mul, sub, and abspow operations into a computation graph.
graph = compose(name="graph")(
    operation(name="mul1", needs=["a", "b"], provides=["ab"])(mul),
    operation(name="sub1", needs=["a", "ab"], provides=["a_minus_ab"])(sub),
    operation(name="abspow1", needs=["a_minus_ab"], provides=["abs_a_minus_ab_cubed"], params={"p": 3})(abspow)
)

pool = Pool(10)
graph.set_execution_method("parallel")
pool.map(lambda i: graph({'a': 2, 'b': 5}, ["a_minus_ab","abs_a_minus_ab_cubed"]), range(100))
```

```
~/Huyng/graphkit/graphkit/network.py in <lambda>(op)
    501
    502             done_iterator = pool.imap_unordered(
--> 503                                 lambda op: (op,op._compute(cache)),
    504                                 upnext)
    505             for op, result in done_iterator:

~/Huyng/graphkit/graphkit/functional.py in _compute(self, named_inputs, outputs)
     17
     18     def _compute(self, named_inputs, outputs=None):
---> 19         inputs = [named_inputs[d] for d in self.needs if not isinstance(d, optional)]
     20
     21         # Find any optional inputs in named_inputs.  Get only the ones that

~/Huyng/graphkit/graphkit/functional.py in <listcomp>(.0)
     17
     18     def _compute(self, named_inputs, outputs=None):
---> 19         inputs = [named_inputs[d] for d in self.needs if not isinstance(d, optional)]
     20
     21         # Find any optional inputs in named_inputs.  Get only the ones that

KeyError: 'a_minus_ab'
```

ankostis commented:
Exactly!

I will work on the tip though, to ensure my changes remain future-proof, and will backport them to 26.

ankostis committed:
FIX(net): iterate_my_self err when droping isolates

ankostis committed:
TEST(plan): new check-multithreading-TC FAILS(!) ...

due to shared `Plan.executed`.

+ fix: enable forgotten check in overrides-TC.

ankostis committed:
FIX(NET,PLAN): MULTITHREAD works with executed LOCAL-VAR, ...

Note that plotting `executed` has not been undone,
it is not receiving ever any set now.

ankostis committed:
doc: number sections, add titles for modifiers

ankostis commented:
Just pushed a fix here for the new "multithreading" TC.

And my mistake on #26, it never shared `executed` variable, so it passes ok with the new "multithreading" TC, you may merge it.

ankostis commented:
For reference i continued the implementation of thi PR in [pygraphkit/graphtik](https://github.com/pygraphkit/graphtik) project, which have departed since, as described in pygraphkit/graphtik#1.
